### ISSUE TITLE
Segmentation fault: 11 with different epoch size

### ISSUE THREAD

**Issue by FabbyD** [facebookresearch/fastText#112]
I encounter a segmentation fault when training word vectors for 5 epochs on a large corpus. I used the same command as the example file (except for number of epochs):
```
./fasttext skipgram -input "${DATADIR}"/mydata.txt -output "${RESULTDIR}"/mydata -lr 0.025 -dim 100 \
  -ws 5 -epoch 5 -minCount 5 -neg 5 -loss ns -bucket 2000000 \
  -minn 3 -maxn 6 -thread 4 -t 1e-4 -lrUpdateRate 100
```
gcc -v output:
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.1.0 (clang-602.0.49) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```
I tried compiling with g++-4.7 instead (replacing the CXX tag in the Makefile) because my clang ~~is~~ was outdated. Same results. (EDIT: I updated XCode to latest version but it still fails.)

Even if my original data is already utf8, I made sure it was by running this command:
`iconv -c -t UTF-8 < mydata.txt > mydata.utf8.txt`

Training for 5 epochs on text9 (the example data) works perfectly fine. Training on my data for 1 epoch is fine, but changing only the number of epochs produces a seg fault. I tried different arguments but it seems to always fail around 4.5%. What else could be wrong with my data? It is a large `.txt` file containing around 136M sentences and 3.6B words. One sentence per line.

**Comment by spate141** [comment]
That seems odd, I did the 500M lines, same parameters, except 16 core and 30GB memory. Works fine for me.

**Comment by EdouardGrave** [comment]
Hi @FabbyD,

Could you please also provide the output of fastText? Similar to:
```
Read 3M words
Number of words:  67051
Number of labels: 0
Progress: 100.0%  words/sec/thread: 41682  lr: 0.000000  loss: 2.952671  eta: 0h0m
```
Thanks!

**Comment by FabbyD** [comment]
Hi @EdouardGrave 

I restarted the program with the same command as above. The output is a little bit weird this time. Instead of clearing the line and printing a new progress, it printed one humongous line with all progress updates one after the other. Here is the output (only the last progress update of course):

```
fabrice$ ./fasttext skipgram -input data/rcv1.wiki.txt -output result/rcv1.wiki.debug -lr 0.025 -dim 100 -ws 5 -epoch 5 -minCount 5 -neg 5 -loss ns -bucket 2000000
Read 3645M words
Number of words:  3813778
Number of labels: 0
Progress: 3.9%  words/sec/thread: 92838  lr: 0.024036  loss: 6.928423  eta: 4h22m Segmentation fault: 11
```

I also updated my compiler.
```
fabrice$ gcc -v
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/c++/4.2.1
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin16.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I'm running it again as I'm writing this just in case.

**Comment by FabbyD** [comment]
@spate141 This machine has 12 cores and 64GB memory. While running, fasttext uses 5.25 GB, I doubt I'm having memory issues. Right? Is there a way to lower memory usage?

**Comment by spate141** [comment]
have you tried with `-loss hs -bucket 1000000` ?

**Comment by FabbyD** [comment]
@spate141 Thank you for the reference! I understand how the architecture works and the different arguments except for number of buckets. It's the only thing I'm not familiar with and, at first glance, that paper doesn't seem to explain what it is.

I did. It actually fails much earlier.
```
fabrice$ ./fasttext skipgram -input data/rcv1.wiki.txt -output result/rcv1.wiki.debug -lr 0.025 -dim 100 -ws 5 -epoch 5 -minCount 5 -loss hs -bucket 1000000
Read 3645M words
Number of words:  3813778
Number of labels: 0
Progress: 0.2%  words/sec/thread: 50428  lr: 0.024938  loss: 82.364296  eta: 8h20m Segmentation fault: 11
```

I would actually like to try with both Hierarchical Softmax and Negative Sampling. I also tried `-loss ns -bucket 1000000`. Same result.

**Comment by spate141** [comment]
@FabbyD I'm sorry, but reducing bucket size is not an optimal thing to do in this case. As I know about how buckets are used in creating model if you can do some experiment with increasing bucket size? 

Here, some information on bucketing and padding.. Its kind of not exactly related but you can get some idea..  https://www.tensorflow.org/versions/r0.12/tutorials/seq2seq/index.html#bucketing-and-padding

**Comment by jamestang0219** [comment]
@FabbyD I encountered same error like yours, and I solve this problem by merging each lines of train   set to one line. Hope it helps for you.

**Comment by FabbyD** [comment]
Like the text9 dataset eh. Wow didn't that would be it. Interesting. I'll try when I get a chance.

**Comment by cpuhrsch** [comment]
I'll close this issue for now as it appears quite old and reasonably resolved. Please feel free to reopen it at any time if the issue persists.
